src/cargo/core/compiler/mod.rs

```diff
@@ -1353,14 +1353,39 @@ fn check_cfg_args(build_runner: &BuildRunner<'_, '_>, unit: &Unit) -> Vec<OsStri
         // Cargo and docs.rs than rustc and docs.rs. In particular, all users of docs.rs use
         // Cargo, but not all users of rustc (like Rust-for-Linux) use docs.rs.
 
-        vec![
+        let mut args = vec![
             OsString::from("--check-cfg"),
             OsString::from("cfg(docsrs)"),
             OsString::from("--check-cfg"),
             arg_feature,
-        ]
+        ];
+
+        // Also include the custom arguments specified in `[lints.rust.unexpected_cfgs.check_cfg]`
+        if let Ok(Some(lints)) = unit.pkg.manifest().resolved_toml().resolved_lints() {
+            if let Some(rust_lints) = lints.get("rust") {
+                if let Some(unexpected_cfgs) = rust_lints.get("unexpected_cfgs") {
```

Comment:
... speaking of inconsistent `-` / `_`.  We didn't do anything to force `-` or `_` in `lints` so you could have `lints.rust.unexpected-cfgs.check-cfg` and it would not be read but instead give an unused key warning

Reply:
See #13943